### PR TITLE
 Backend Comments Test Cases

### LINKED ISSUE #44: Backend Comments Need Test cases
**Description of the issue**
The test.py in the backend Comments folder does not have any test cases, that should be used to test the other files in the folder.

**Files that need to be updated**
test.py

**Link**
#39 

**Screenshots/Example (Optional)**
q3
Here's a view of the file

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by misherwa**
*Please complete the checklist so the group can quickly understand your PR.*

### Category
- [x] **Feature**
- [ ] **Bug Fix**
- [x] **Testing**
- [ ] **Documentation**
- [ ] **Other**

### Domain
- [ ] **Frontend**
- [x] **Backend**
- [ ] **Both**

### Description
Provide test case coverage for all the files in the Comments backend folder. I have completed test coverage for models.py, serializers.py, and views, py. I've also added error handling and authorization requirements features. In views.py, a look_up field was added to retrieve a certain comment, and deleting can be done by the author, moderator, or hub owner.

### Related Issue(s)
#44 "Backend Comments Need Test Cases"

### New Additions
No new files

### Modifications
-tests.py (Added test coverage for models.py)
-urls.py (Removed a copy line for commentdelete)
-serializers.py (Added file changes from PR #67 and error handling)
-views.py (Added authorization and error handling)

### New Dependencies
None

### Visuals
q4
Here is a current view of the coverage results for the Comments folder with 93 percent coverage. 

### Architecture of Your Changes
Made authorization checks in views.py for the CommentDelete class. This is to ensure that another user can not delete a comment. Authorization checks for replying to liking or disliking replies, which should not be possible since the feature has not been implemented to replies. Liking and disliking are currently only for comments or posts.

No other modifications to an existing system, only improvements to test coverage results and error handling.

### Usage and Integration
Just used to increase our overall test coverage percentage,

### Testing
Describe the tests performed to ensure your changes function as expected. This section should include:
- **Testing Approach**: Applied unit testing 
- **Edge Cases**: Currently need to add coverage for test cases that involve the serializers.py and views.py files in the Comments backend folder.
- **Test Commands**: Make sure to do "pip install coverage", then in the backend folder do "python manage.py test". Lastly, you can do "coverage report" for the result

### Mentions
N/A

### CODE REVIEW

**Comment by thecatfour** [comment]
I don't think "test_create_comment" is the best name for this test case. The name implies that the test is creating a comment, but the comment is created in the setUp function. Something like "test_check_setup_comment" might be more accurate.

image

**Review by leed85 — COMMENTED** [review]
Good tests, I think we need to discuss the architecture of comments a bit, should we allow constant reply of comments(i.e. reddit style where you can reply to a specific comment or a reply and it constantly branches off from there) or like youtube where you just @ the person in the comments as a reply

**Comment by misherwa** [comment]
> test_check_setup_comment

Yeah that makes sense, I changed "test_create_comment" to "test_check_setup_comment"

**Comment by misherwa** [comment]
> Good tests, I think we need to discuss the architecture of comments a bit, should we allow constant reply of comments(i.e. reddit style where you can reply to a specific comment or a reply and it constantly branches off from there) or like youtube where you just @ the person in the comments as a reply

I believe it should be more like YouTube. It's more of a personal preference for me, but I think it would be safer to implement for now and we can try testing out a branching method,

**Comment by misherwa** [comment]
I modified the "to_representation" and "validate" methods in the  seralizers.py file. I added error handling for both. 

The error handling for "to_representation" can be found in the if statement right before "representation" gets returned. It is also in the CommentSerializer class.

"validate" is in the "CommentCreateSerializer" class, which includes "try and except" when issues if a comment or post doesn't exist

**Comment by thecatfour** [comment]
Is there a test for checking if a hub owner/mod can delete a comment on a post in their hub? I can only find 2 delete tests that involve a non-author trying to delete a comment and the author trying to delete a comment. I think its important to make sure mods and owners can delete comments, so I think you should add those cases.

**Comment by misherwa** [comment]
> Is there a test for checking if a hub owner/mod can delete a comment on a post in their hub? I can only find 2 delete tests that involve a non-author trying to delete a comment and the author trying to delete a comment. I think its important to make sure mods and owners can delete comments, so I think you should add those cases.

Gotcha, I updated the code to include test cases for when a hub owner or mod wants to delete a comment:
org


Note: I had to update the views.py file for Comments to include cases if a hub owner or mod wants to delete a comment. I also added a look_up field and set it to "comment_id".  It's similar  to how it's done in for views.py in Posts, so now the "CommentDelete" class looks like this:
cur

**Comment by thecatfour** [comment]
I'm not sure if including the .coverage file in the PR is necessary. It might cause conflicts similar to the ones we got earlier when trying to merge with the database file. Other than that, I think this includes all the tests that we need for the comments right now. If we include a view for editing a comment, we will need to add test cases for that though.

**Comment by nnguyen144** [comment]
> > Good tests, I think we need to discuss the architecture of comments a bit, should we allow constant reply of comments(i.e. reddit style where you can reply to a specific comment or a reply and it constantly branches off from there) or like youtube where you just @ the person in the comments as a reply
> 
> I believe it should be more like YouTube. It's more of a personal preference for me, but I think it would be safer to implement for now and we can try testing out a branching method,

YouTube comment styling is easier and just requires a link to the post and to another comment if the comment is a reply. I believe YouTube comment styling would be the way to go.

File: backend/Comments/tests.py
Comment (nnguyen144) [inline backend/Comments/tests.py]: Should be 'other_user_password' and not 'other_user_passwird'

File: backend/Comments/tests.py
Comment (nnguyen144) [inline backend/Comments/tests.py]: Small nitpick but you could probably de-capitalize Moderator and Hub in the titles of these two tests

**Review by nnguyen144 — CHANGES_REQUESTED** [review]
I also noticed that you reversed a lot of urls that are not in urls.py, such as 
url = reverse("comments:like-comment", kwargs={"comment_id": reply.id})
Make sure you fix this.

**Comment by nnguyen144** [comment]
Also since you added a new URL, can you add that URl to url-segments.ts

**Comment by thecatfour** [comment]
> Also since you added a new URL, can you add that URl to url-segments.ts

I already implemented the new URLs in #67, so it should be the same whether they are added here or in that PR.

**Review by thecatfour — APPROVED** [review]
I believe this addresses what we need for the backend comments tests.

**Comment by misherwa** [comment]
> I believe this addresses what we need for the backend comments tests.

I think I'll wait until your #67 PR gets merged before making proper changes to this one. But for now, I'll wait until your PR is ready to be reviewed and I can take a look at it. Or I can take a look at your draft if now if you think I should.

**Review by nnguyen144 — APPROVED** [review]

**Pull request closed without merging**
